app/src/main/java/com/github/libretube/services/PlaylistDownloadEnqueueService.kt

```diff
@@ -0,0 +1,194 @@
+
+            if (playlistPage == null && alreadyRetriedOnce) {
+                toastFromMainDispatcher(R.string.unknown_error)
+                onDestroy()
+                return
+            }
+
+            if (playlistPage == null) {
+                // retry if previous attempt failed
+                alreadyRetriedOnce = true
+                continue
+            }
+
+            alreadyRetriedOnce = false
+            enqueueStreams(playlistPage.relatedStreams)
+            nextPage = playlistPage.nextpage
+        }
+    }
+
+    private suspend fun enqueueStreams(streams: List<StreamItem>) {
+        withContext(Dispatchers.Main) {
+            nManager.notify(ENQUEUE_PROGRESS_NOTIFICATION_ID, buildNotification())
+        }
+
+        for (stream in streams) {
+            val videoInfo = runCatching {
+                RetrofitInstance.api.getStreams(stream.url!!.toID())
+            }.getOrNull() ?: continue
+
+            val videoStream = getStream(videoInfo.videoStreams, maxVideoQuality)
+            val audioStream = getStream(videoInfo.audioStreams, maxAudioQuality)
+
+            val downloadData = DownloadData(
+                videoId = stream.url!!.toID(),
+                fileName = videoInfo.title,
+                videoFormat = videoStream?.format,
+                videoQuality = videoStream?.quality,
+                audioFormat = audioStream?.format,
+                audioQuality = audioStream?.quality,
+                audioLanguage = audioLanguage.takeIf {
+                    videoInfo.audioStreams.any { it.audioTrackLocale == audioLanguage }
+                },
+                subtitleCode = captionLanguage.takeIf {
+                    videoInfo.subtitles.any { it.code == captionLanguage }
+                }
+            )
+            DownloadHelper.startDownloadService(this, downloadData)
+
+            amountOfVideosDone++
+            withContext(Dispatchers.Main) {
+                nManager.notify(ENQUEUE_PROGRESS_NOTIFICATION_ID, buildNotification())
+            }
+        }
+
+        if (amountOfVideos == amountOfVideosDone) onDestroy()
+    }
+
+    private fun getStream(streams: List<PipedStream>, maxQuality: Int?): PipedStream? {
+        val maxStreamQuality = maxQuality ?: return null
+
+        // sort streams by their quality/bitrate
+        val sortedStreams = streams
+            .sortedBy { it.quality.getWhileDigit() }
+
+        // return the last item below the maximum quality - or if there's none - the stream with
+        // the lowest quality available
+        return sortedStreams
+            .lastOrNull { it.quality.getWhileDigit()!! <= maxStreamQuality }
+            ?: sortedStreams.firstOrNull()
+    }
+
+    override fun onDestroy() {
+        ServiceCompat.stopForeground(this, ServiceCompat.STOP_FOREGROUND_REMOVE)
+        stopSelf()
+
+        super.onDestroy()
+    }
+
+    companion object {
+        private const val ENQUEUE_PROGRESS_NOTIFICATION_ID = 3
```

Comment:
Rather than having this here as a magic number, would it make sense to create an enum somewhere else along with other notification IDs? It's difficult to understand why 3 is used otherwise. (I'm guessing 1 and 2 are used elsewhere in the app)

This could be another PR perhaps.

Reply:
Yes, we should probably do thay in an other PR. 1 and 2 are used for the player and the normal download notification.